core/src/ledger_cleanup_service.rs

```diff
@@ -286,6 +307,12 @@ mod tests {
         assert!(shreds_per_slot > 1);
         blockstore.insert_shreds(shreds, None, false).unwrap();
 
+        // Explicitly drop and re-open Blockstore to force a flush as the method
+        // used in LedgerCleanupService::find_slots_to_clean(() doesn't find
+        // shreds in memtables, only in SST's
+        drop(blockstore);
+        let blockstore = Arc::new(Blockstore::open(ledger_path.path()).unwrap());
```

Comment:
This is pretty dirty, but not sure of how else to address the issue I called out in the comment

Reply:
I think it's because `live_files_metadata` only reads file metadata and all the in-memory data will be omitted.

Two alternatives in my mind at this moment:

* Do a manual flush(), or
* Do the iterator as we did in the past but narrow the iteration using the return value of the new `find_slots_to_clean`.

I will think more about it and see whether we can find a better solution.

Reply:
> Do a manual flush(), or

Ahh, didn't realize this was an option. I'm in favor of doing this approach over my hack if we end up going that route. Should be just a couple wrappers that I'll add in another PR.

> Do the iterator as we did in the past but narrow the iteration using the return value of the new

I'm assuming the idea here would be to to figure out the max slot that we saw from scanning all SST's, and then do `SlotMeta` iterator past that?

This would be more accurate, but I also kind of like not having the scan the iterator at all. Given that this service's goal is to keep disk usage in check, I think it is fair to only consider the data on disk and ignore the stuff in memory.

That being said, I now realize that my use of `Blockstore::highest_slot()` is likely returning a slot that is in the memtable. When I initially coded this up, I was decoding the keys from the SST's and tracking min/max; I think I may go back to that approach. Doing so would be more consistent with only cleaning up data that has already been persisted.

Reply:
Wait.  If things are still in memtable, does it mean they are relatively new so that they don't need to be cleaned up?

If that's the case, then we shouldn't need the information from memtable.

Am I right at this part?

Reply:
> Wait. If things are still in memtable, does it mean they are relatively new 

Yes, I believe this to be accurate

> so that they don't need to be cleaned up?

The data in the memtables will not be cleaned up itself, but I guess the question we're hitting is whether we should count the memtable data when figuring out how many shreds we have and if/how many shreds we need to clean.

> If that's the case, then we shouldn't need the information from memtable.

This is the way I am leaning as well. Going back to my earlier comment, `LedgerCleanupService` is meant to keep disk space in check. We do so by counting shreds for simplicity, but we really care about is the footprint on disk. Given that memtables do not contribute to disk usage (at the time that they are still memtables), I think not counting the memtable data is fair from idealogical standpoint.

From a practical standpoint, not counting the memtables keeps the actual code simpler / cleaner while adding a little extra burden (having to flush) to test code. I think this is a good trade to make

Reply:
Hmmm, calling both [`flush_opt()`](https://docs.rs/rocksdb/latest/rocksdb/struct.DBCommon.html#method.flush_opt) and [`flush_wal()`](https://docs.rs/rocksdb/latest/rocksdb/struct.DBCommon.html#method.flush_wal) didn't help; I tried both orders and saw in the LOG file that the manual flush was getting initiated. However, no SST files were created.

I'll have to dig around for this, but I seem to remember the WAL getting "processed" when you open the DB; maybe this is the bit that is getting SST's created?

What are you thoughts on me writing a helper function in `ledger_cleanup_service` test scope that just closes and opens the DB? I kind of don't want to spend more time getting to the bottom of why SST's aren't getting created, especially since we only need this behavior for these unit tests

Reply:
Let's continue in:
https://github.com/solana-labs/solana/pull/29239#discussion_r1050431226

This thread already kind of long and we discussed several things